openedx/core/djangoapps/catalog/utils.py

```diff
@@ -239,6 +242,47 @@ def get_course_runs_for_course(course_uuid):
     else:
         return []
 
+def get_visible_course_runs_for_course(entitlement, course_uuid):
+    """
+    We only want to show courses for a particular entitlement that:
+
+    1) Are currently running or in the future
+    2) A user can enroll in
+    3) A user can upgrade in
+    4) Are published
+    """
+    course_runs_for_course = get_course_runs_for_course(course_uuid)
+    enrollable_course_runs = []
+
+    # Only show published course runs that can still be enrolled and upgraded
+    now = datetime.datetime.now(UTC)
+    for course_run in course_runs_for_course:
+        # Only courses that have not ended will be displayed
+        run_start = course_run.get('start')
+        run_end = course_run.get('end')
+        is_running = run_start and run_end and datetime_parse(run_end) > now
+
+        # Only courses that can currently be enrolled in will be displayed
+        enrollment_start = course_run.get('enrollment_start')
+        enrollment_end = course_run.get('enrollment_end')
+        can_enroll = ((not enrollment_start or datetime_parse(enrollment_start) < now)
```

Comment:
For clarity move the datetime_parse up to the line where you get the enrollment_start, enrollment_end, run_start, and run_end.  It will make the rest of the code easier to follow I think.

`enrollment_start = datetime_parse(enrollment_start)`

Reply:
of course if you're going to do that you should change
`enrollment_end = course_run.get('enrollment_end')`
to
`enrollment_end = datetime_parse(enrollment_start)`
as well

Reply:
These both throw errors if run_start or run_end don't exist and the logic, and the None checkers make the code a lot messier, going to leave as is.